toolchain/lex/tokenized_buffer.h

```diff
@@ -260,55 +260,101 @@ class TokenizedBuffer : public Printable<TokenizedBuffer> {
     int indent;
   };
 
-  struct TokenInfo {
-    TokenKind kind;
+  class TokenInfo {
+   public:
+    auto kind() const -> TokenKind { return TokenKind::FromInt(kind_); }
+    auto has_leading_space() const -> bool { return has_leading_space_; }
+
+    auto ident_id() const -> IdentifierId {
+      CARBON_DCHECK(kind() == TokenKind::Identifier);
+      return IdentifierId(token_payload_);
+    }
+    auto set_ident_id(IdentifierId ident_id) -> void {
+      CARBON_DCHECK(kind() == TokenKind::Identifier);
+      CARBON_DCHECK(ident_id.index < (2 << PayloadBits));
+      token_payload_ = ident_id.index;
+    }
+
+    auto string_literal_id() const -> StringLiteralValueId {
+      CARBON_DCHECK(kind() == TokenKind::StringLiteral);
+      return StringLiteralValueId(token_payload_);
+    }
+
+    auto int_id() const -> IntId {
+      CARBON_DCHECK(kind() == TokenKind::IntLiteral ||
+                    kind() == TokenKind::IntTypeLiteral ||
+                    kind() == TokenKind::UnsignedIntTypeLiteral ||
+                    kind() == TokenKind::FloatTypeLiteral);
+      return IntId(token_payload_);
+    }
+
+    auto real_id() const -> RealId {
+      CARBON_DCHECK(kind() == TokenKind::RealLiteral);
+      return RealId(token_payload_);
+    }
+
+    auto closing_token_index() const -> TokenIndex {
+      CARBON_DCHECK(kind().is_opening_symbol());
+      return TokenIndex(token_payload_);
+    }
+    auto set_closing_token_index(TokenIndex closing_index) -> void {
+      CARBON_DCHECK(kind().is_opening_symbol());
+      CARBON_DCHECK(closing_index.index < (2 << PayloadBits));
+      token_payload_ = closing_index.index;
+    }
+
+    auto opening_token_index() const -> TokenIndex {
+      CARBON_DCHECK(kind().is_closing_symbol());
+      return TokenIndex(token_payload_);
+    }
+    auto set_opening_token_index(TokenIndex opening_index) -> void {
+      CARBON_DCHECK(kind().is_closing_symbol());
+      CARBON_DCHECK(opening_index.index < (2 << PayloadBits));
+      token_payload_ = opening_index.index;
+    }
+
+    auto error_length() const -> int {
+      CARBON_DCHECK(kind() == TokenKind::Error);
+      return token_payload_;
+    }
+
+    auto byte_offset() const -> int32_t { return byte_offset_; }
+
+    TokenInfo(TokenKind kind, bool has_leading_space, int32_t byte_offset)
+        : kind_(kind.AsInt()),
+          has_leading_space_(has_leading_space),
+          byte_offset_(byte_offset) {}
+
+    TokenInfo(TokenKind kind, bool has_leading_space, int payload,
+              int32_t byte_offset)
+        : kind_(kind.AsInt()),
+          has_leading_space_(has_leading_space),
+          token_payload_(payload),
+          byte_offset_(byte_offset) {
+      CARBON_DCHECK(payload >= 0 && payload < (2 << PayloadBits));
+    }
+
+   private:
+    static constexpr int PayloadBits = 23;
```

Comment:
Words outside of comments? wow. Can you send me a reference to this?

My generated 256k line file, which is *huge*, still isn't that big.

This would be a file that *at it's smallest* is 150mb of source code text... And much more likely those words average out to 3-4 bytes per word, so we're talking about more like 250mb of source code text.... Maybe we could insist on generated code splitting into multiple files at that point?

Reply:
"Words" is as defined by `wc`, so it includes comment words. And yes, it's over 850 Mb.

Reply:
(For others, Geoff was able to point me at the specific example.)

While the proposed scheme would preclude this, I think that's a reasonable implementation limit to impose in order to make essentially all other code more efficient.

The code in question is generated code for protocol buffers, where the `.proto` file itself is *also* generated and nearly 9mb long.

I think it would be entirely reasonable to say this isn't supported and insist that users desiring something of this form generate multiple files.

That said, if we in practice find this to be a difficult limitation, I think we can easily remove the limit in question on the number of tokens: instead of storing a token index directly here, we can store an index into a table of open/close pairs -- we can even shard that into separate tables for each kind of grouping symbol. This would allow `2^22 - 1` tokens with inline grouping indices, and if more tokens are found, fall back to separate tables for each kind, and only run out at `2^22 - 1` pairs of a single grouping token. That would be a file with 4mb of curlies or parens alone, and in a single file with no `#include`s. And if we want to go still further, we could begin dropping the cross-reference index at that point and degrade diagnostics instead.

Overall, this seems like a plausible limit to try to hold initially and one we could credibly relax in tactical ways if specific problems arise that we can't reasonably ask users to work around.